### PR TITLE
Introduce support for local archived binary targets

### PR DESCRIPTION
**Pull request by dmhts**
This PR introduces support for local archived binary targets.

### Motivation:

In order to significantly speed up builds and make them more reliable (especially in a CI environment) it makes sense to precompile and put 3rd party dependencies under source control to later consume them as local binary targets.
To facilitate the managing and maintaining of those binary targets it would make sense to consume them in an archived form because that would decrease the number of files and the general repository size by a lot (sometimes by orders of magnitude), as well as present cleaner diffs while updating versions of the targets.
To address the issues mentioned above we need to introduce support for local archived binary targets to SwiftPM. This is the main motivation for this PR.

The initial discussion of the feature started [here](https://forums.swift.org/t/supporting-zip-archives-for-local-binary-targets/50873).

A big thanks to @Mo7amedFouad for the initial work that brought us to the actual use case and the following idea of local archives!

Note: 
My assumption is that we do not need to ask the client to manually specify a checksum for local archived binary targets, since the artifacts are already on the local machine; such checksum specification will, therefore, not really help with identifying security risks. We may want to automatically generate some kind of a file signature (either a checksum or modification time) to optimize the reusing of already existing managed artifacts (more on this below in the code comments).

### Modifications:

- `ManifestLoader`: local targets now support the new zip extension
- `PackageBuilder`: binary artifact pathfinding logic is handled consistently now, so it covers `local (archived)` and `remote` cases in the same condition. 
- `Workspace`:
    - adjust `func updateBinaryArtifacts(...)` to handle local archived artifacts and all the possible artifact source (`local`, `local-archived` and `remote`) transition permutations.
    - introduce `func extract(_ artifacts: [ManagedArtifact], ...) -> [ManagedArtifact]` to extract local archives.
- `WorkspaceTests`: add tests to cover all (un)happy scenarios and edge cases we can currently think of
- `PD_5_3_LoadingTests`: ensures that the 5.3+ manifest versions support the new binary target path extension
- `Workspace/Diagnostics`: introduce the new local artifact diagnostic errors

### Result:

```swift
let package = Package(
    name: "Dependencies",
    ...
    targets: [
        .binaryTarget(
            name: "TargetA",
            path: "XCFrameworks/TargetA.zip"          
        ),
        .binaryTarget(
            name: "TargetB",
            path: "XCFrameworks/DoesNotReallyMatchTargetBName.zip"
        ),
        .binaryTarget(
            name: "TargetC",
            path: "ArtifactBundles/TargetC.zip"
        ),
        ...
    ]
)
```

### CODE REVIEW

File: Sources/Workspace/Workspace.swift [hunk 1]
```diff
@@ -1707,30 +1714,60 @@ extension Workspace {
...
+
+            if let existingArtifact = existingArtifact {
+                switch existingArtifact.source {
+                case .remote:
+                    artifactsToRemove.append(existingArtifact)
+                case .localArchived:
+                    // TODO: It is better to compare the file signatures of the archives to optimize the
+                    // performance by reusing the existing artifact instead of re-extracting the old one
+                    // each time.
+                    artifactsToRemove.append(existingArtifact)
```
Comment (dmhts) [inline Sources/Workspace/Workspace.swift:1756]: Here we may want to reuse the already existing managed artifact by additionally persisting its file signature and comparing it with the potential artifact (just as it is currently implemented for the remote artifacts, but with the automatically generated signature).

File: Sources/Workspace/Workspace.swift [hunk 2]
```diff
@@ -1975,6 +2034,74 @@ extension Workspace {
 
         return result.map{ $0 }
     }
+    
+    private func extract(_ artifacts: [LocalArchivedArtifact], diagnostics: DiagnosticsEngine) -> [ManagedArtifact] {
```
Comment (dmhts) [inline Sources/Workspace/Workspace.swift:2038]: The DRY principle is not completely followed here since the function copies a huge chunk of the extraction logic from `func download(_ artifacts: [RemoteArtifact] ...) -> [ManagedArtifact]`, I went this way to minimize the amount of refactoring in the PR, but if it is critical I could introduce a more generic function e.g. `func extract(_ artifacts: [UnmanagedArtifact] ...) -> [ManagedArtifacts]`, a price for that would be having more complicated logic and a few additional types, or come up with something else.

File: Sources/Workspace/WorkspaceState.swift [hunk 3]
```diff
@@ -265,6 +265,9 @@ fileprivate struct WorkspaceStateStorage {
                     switch kind {
                     case "local":
                         self.init(underlying: .local)
+                    case "localArchived":
```
Comment (dmhts) [inline Sources/Workspace/WorkspaceState.swift:268]: I don't see any tests for `WorkspaceState`? Is it because a DiagnosticsEngine is not yet available in this part, or should I introduce ones?

**Comment by tomerd** [comment]
thanks for the PR @dmhts , will review it shortly!

File: Sources/Workspace/ManagedArtifact.swift [hunk 4]
```diff
@@ -70,6 +80,21 @@ extension Workspace {
...
+            path: AbsolutePath,
+            archivePath: AbsolutePath
+        ) -> ManagedArtifact {
+            return ManagedArtifact(
+                packageRef: packageRef,
+                targetName: targetName,
+                source: .localArchived(archivePath: archivePath),
+                path: path
+            )
+        }
```
Comment (tomerd) [inline Sources/Workspace/ManagedArtifact.swift:97]: could removing the special type for archived local artifact simplify the solution?

For example, if we do not add `localArchived` and instead allow `local` to take in a directory or a zip file, then in the workspace code where we handle the artifacts we would check for if the artifact path is a directory or a zip file we could probably achieve the same with less impact?

File: Sources/Workspace/Workspace.swift [hunk 5]
```diff
@@ -1768,28 +1809,46 @@ extension Workspace {
...
-                            path: absolutePath)
-                    )
+                    
+                    if absolutePath.extension == LocalArchivedArtifact.extension {
+                        localArchivedArtifacts.append(
+                            LocalArchivedArtifact(
+                                packageRef: packageReference,
+                                targetName: target.name,
+                                archivePath: absolutePath,
+                                destinationPath: self.location.artifactsDirectory.appending(components: packageReference.name))
```
Comment (tomerd) [inline Sources/Workspace/Workspace.swift:1839]: in the local case we set the path to absolutePath which is computed off the relative path provided by the user in the manifest, in effort to reduce the API changes, maybe we can compute the destinationPath from that path also, i.e. 
`self.location.artifactsDirectory.appending(RelativePath(validating: path)` or something along these lines which can be computed and should be unique.

File: Sources/Workspace/Workspace.swift (hunk 2 above)
Comment (tomerd) [reply Sources/Workspace/Workspace.swift:2041]: I think this is fine for now, once the PR is more mature we can see if we can make this DRYer

File: Sources/Workspace/ManagedArtifact.swift (hunk 4 above)
Comment (dmhts) [reply Sources/Workspace/ManagedArtifact.swift:97]: Thanks a lot for the feedback, now when I know the additional constraints (minimizing the API impact), I'll try to optimize the solution for that.

File: Sources/Workspace/Workspace.swift (hunk 5 above)
Comment (dmhts) [reply Sources/Workspace/Workspace.swift]: Done, now `updateBinaryArtifacts` is the only function that underwent changes in `Wokrspace`.

File: Sources/Workspace/WorkspaceState.swift (hunk 3 above)
Comment (dmhts) [reply Sources/Workspace/WorkspaceState.swift]: It is not needed anymore.

File: Sources/Workspace/ManagedArtifact.swift (hunk 4 above)
Comment (dmhts) [reply Sources/Workspace/ManagedArtifact.swift]: Done, there are zero API changes now, and only the private implementations are affected.

File: Sources/Workspace/Workspace.swift (hunk 1 above)
Comment (dmhts) [reply Sources/Workspace/Workspace.swift]: It points out the outdated line now.

File: Sources/Workspace/Workspace.swift [hunk 6]
```diff
@@ -1696,12 +1697,18 @@ extension Workspace {
                 targetName: artifact.targetName
             ]
 
-            if let existingArtifact = existingArtifact, case .remote = existingArtifact.source {
-                // If we go from a remote to a local artifact, we can remove the old remote artifact.
-                artifactsToRemove.append(existingArtifact)
+            if artifact.path.extension == "zip" {
```
Comment (dmhts) [inline Sources/Workspace/Workspace.swift:1700]: The "zip" string literal is already used a few times through the codebase, would it make sense to introduce a type/instance property (e.g. `supportedArchiveType/Extension`) to `ManifestLoader`, for example (but that would be a public API change)?

File: Sources/Workspace/Workspace.swift [hunk 7]
```diff
@@ -1696,12 +1697,18 @@ extension Workspace {
...
+                // TODO: We could check here a file signature of the `existingArtifact`, and
+                // if it matches the manifest `artifact` signature we could just `continue`.
+                artifactsToExtract.append(artifact)
```
Comment (dmhts) [inline Sources/Workspace/Workspace.swift:1703]: If we want to address the `todo:` that most probably would require persisting the `existingArtifact` file signature (either a checksum or modification date) in the workspace state.

File: Sources/PackageLoading/PackageBuilder.swift [hunk 8]
```diff
@@ -768,7 +768,9 @@ public final class PackageBuilder {
                 providers: manifestTarget.providers
             )
         } else if potentialModule.type == .binary {
-            guard let artifact = self.binaryArtifacts.first(where: { $0.path == potentialModule.path }) else {
+            guard let artifact = self.binaryArtifacts.first(where: {
+                $0.path == potentialModule.path || $0.path.basenameWithoutExt == potentialModule.path.basenameWithoutExt
```
Comment (dmhts) [inline Sources/PackageLoading/PackageBuilder.swift:772]: With the current approach, the base name is the only thing that we can compare to match the artifacts (and that is the price for keeping the public API intact), and that means that the local archive base name must always match the target name. I'm wondering if that is a reasonable restriction (which we would have to reflect in the documentation), or should we try to do something about that?

Comment (tomerd) [reply Sources/PackageLoading/PackageBuilder.swift:772]: would the whole path be the same, other than the extension?

File: Sources/Workspace/Workspace.swift (hunk 6 above)
Comment (tomerd) [reply Sources/Workspace/Workspace.swift:1700]: thats great yes

File: Sources/Workspace/Workspace.swift (hunk 7 above)
Comment (tomerd) [reply Sources/Workspace/Workspace.swift:1703]: adding that to workspace artifacts state makes sense to me

**Review by tomerd — COMMENTED** [review]
this is coming together nicely, couple of ideas inline

File: Sources/PackageLoading/PackageBuilder.swift (hunk 8 above)
Comment (dmhts) [reply Sources/PackageLoading/PackageBuilder.swift]: That's a great question, it turned out that the paths were different because the potential module path was determined incorrectly at the earlier step. It is fixed now by handling local archives as a special case of manifest targets with custom paths: https://github.com/apple/swift-package-manager/pull/3746/files#diff-68f661bfe880d4bf7059718964ee104add3839e7e2f5bc8bfd644ab706ff7149R551-R556

File: Sources/PackageLoading/PackageBuilder.swift [hunk 9]
```diff
@@ -548,13 +548,12 @@ public final class PackageBuilder {
                 if fileSystem.isDirectory(path) {
                     return path
                 }
+                if fileSystem.isFile(path) && path.extension == "zip" {
```
Comment (tomerd) [inline Sources/PackageLoading/PackageBuilder.swift:551]: is the target.type == .binary  in this case too?

Comment (dmhts) [reply Sources/PackageLoading/PackageBuilder.swift]: Right, this check is not strict enough, since the other target types potentially may meet these criteria, fixed that by adding a check for the `.binary` type (also updated the `binaryArtifacts` array description, since it contains not only downloaded artifacts). 
Unfortunately, I don't see how to make it more readable, e.g. by introducing the `target.isLocalArchivedBinary` predicate, but maybe it is already good enough.

I believe this is the right place to look for the matching managed artifact path, just right after the manifest target path is validated. The only thing that I find strange, in the case of binary targets the path validation is performed when the managed artifacts are already in place.

Comment (tomerd) [reply Sources/PackageLoading/PackageBuilder.swift]: I wonder if we flip the order of the if here it would help make this cleaner?

Comment (dmhts) [reply Sources/PackageLoading/PackageBuilder.swift]: That's [what I did](https://github.com/apple/swift-package-manager/pull/3746/commits/b748d4808aad9fd59cc3a2ae874dd55673abfbfa) in the beginning :) but then reverted the change assuming that local binary targets must first validate their path. But now having more context and knowing that [binary target path validation](https://github.com/apple/swift-package-manager/blob/3d48f94174fe484afbb854abb806cac9a51634ed/Sources/Workspace/Workspace.swift#L1805) is performed before the one in `PackageBuilder` I think we can safely flip the order (thanks to @neonichu [for fixing the missing validation crash](https://github.com/apple/swift-package-manager/pull/3737)). 
That actually makes the logic more consistent and clear, emphasizing that all binary target types do not need the additional path validation at the `PackageBuilder` level.
[Done.](https://github.com/apple/swift-package-manager/pull/3746/commits/2fb30110e2d93937b2d8c4c2c52471658b0597be)

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by dmhts** [comment]
@swift-ci please smoke test

File: Sources/PackageLoading/PackageBuilder.swift (hunk 9 above)
Comment (dmhts) [reply Sources/PackageLoading/PackageBuilder.swift]: Just merged the `main` branch in and adapted to the new API, you could try to run the smoke tests again.

**Comment by tomerd** [comment]
@swift-ci please smoke test

File: Sources/Workspace/Workspace.swift (hunk 7 above)
Comment (dmhts) [reply Sources/Workspace/Workspace.swift:1730]: [Done](https://github.com/apple/swift-package-manager/pull/3746/commits/4f581685bb840a8f7a2973f364ce8350db779687), I introduced the checksum as the local source case associated value. On the one hand, I like the symmetry with the remote source case, but on the other, the optional checksum requires some additional handling. Would it work for you like that, or would you prefer to iterate on that by storing the checksum in a separate `localArchived` case, renaming the associated value name to `archiveChecksum`, or maybe something else?

File: Sources/PackageLoading/PackageBuilder.swift [hunk 10]
```diff
@@ -529,8 +529,13 @@ public final class PackageBuilder {
...
         /// Returns the path of the given target.
         func findPath(for target: TargetDescription) throws -> AbsolutePath {
-            // If there is a custom path defined, use that.
-            if let subpath = target.path {
+            if target.type == .binary {
+                if let artifact = self.binaryArtifacts.first(where: { $0.path.basenameWithoutExt == target.name }) {
+                    return artifact.path
+                } else {
+                    throw ModuleError.artifactNotFound(target.name)
+                }
```
Comment (tomerd) [inline Sources/PackageLoading/PackageBuilder.swift:537]: nit:

```
guard let artifact = self.binaryArtifacts.first(where: { $0.path.basenameWithoutExt == target.name }) else {
  throw ModuleError.artifactNotFound(target.name)
}
return artifact.path
```

File: Sources/SPMTestSupport/MockWorkspace.swift [hunk 11]
```diff
@@ -544,8 +544,8 @@ public final class MockWorkspace {
             case (.remote(let lhsURL, let lhsChecksum), .remote(let rhsURL, let rhsChecksum)):
                 XCTAssertEqual(lhsURL, rhsURL, file: file, line: line)
                 XCTAssertEqual(lhsChecksum, rhsChecksum, file: file, line: line)
-            case (.local, .local):
-                break
+            case (.local(let lhsChecksum), .local(let rhsChecksum)):
+                XCTAssertEqual(lhsChecksum, rhsChecksum, file: file, line: line)
```
Comment (tomerd) [inline Sources/SPMTestSupport/MockWorkspace.swift:548]: <3

File: Sources/Workspace/Workspace.swift (hunk 7 above)
Comment (tomerd) [reply Sources/Workspace/Workspace.swift:1730]: I think this is good

File: Sources/Workspace/Workspace.swift [hunk 12]
```diff
@@ -1718,12 +1719,23 @@ extension Workspace {
...
-            if let existingArtifact = existingArtifact, case .remote = existingArtifact.source {
-                // If we go from a remote to a local artifact, we can remove the old remote artifact.
-                artifactsToRemove.append(existingArtifact)
+            if artifact.path.extension == "zip" {
+                // If we already have an artifact that was extracted from an archive with the same checksum,
+                // we don't need to extract the artifact again.
+                if case .local(let existingChecksum) = existingArtifact?.source,
+                   existingChecksum == self.checksum(forBinaryArtifactAt: artifact.path, diagnostics: diagnostics) {
+                    continue
+                }
```
Comment (tomerd) [inline Sources/Workspace/Workspace.swift:1728]: formatting ^^

File: Sources/Workspace/Workspace.swift [hunk 13]
```diff
@@ -1732,19 +1744,24 @@ extension Workspace {
...
+                    // If the checksum is different but the package wasn't updated, this is a security risk.
+                    if !addedOrUpdatedPackages.contains(artifact.packageRef) {
+                        diagnostics.emit(.artifactChecksumChanged(targetName: artifact.targetName))
+                        continue
+                    }
                 }
 
-                artifactsToRemove.append(existingArtifact)
+                if isAtArtifactsDirectory(existingArtifact) {
+                    // Remove the old extracted artifact, be it local archived or remote one.
```
Comment (tomerd) [inline Sources/Workspace/Workspace.swift:1762]: we should make sure this change does not remove local (non-archived) directories, which would delete user's code. if there isn't a unit test for this already, worth adding one

File: Sources/Workspace/Workspace.swift [hunk 14]
```diff
@@ -2000,6 +2021,79 @@ extension Workspace {
...
+                }
+            })
+        }
+
+        group.wait()
+
+        return result.map{ $0 }
+    }
+
+    private func isAtArtifactsDirectory(_ artifact: ManagedArtifact) -> Bool {
```
Comment (tomerd) [inline Sources/Workspace/Workspace.swift:2094]: nit: this can be a private extension on ManagedArtifact

File: Sources/Workspace/Workspace.swift (hunk 12 above)
Comment (dmhts) [reply Sources/Workspace/Workspace.swift:1729]: `SwiftFormat` [moves the bracket to the next line](https://github.com/apple/swift-package-manager/pull/3746/commits/5e2d7bbe8937164b5b8cadfcdd15d4066c423886), unfortunately, it is very inconvenient to run it since it formats a lot of not affected by the PR code (so I have to discard a lot of not relevant changes each time), what would be the right way to check formatting?
```
swiftformat .                                                                           
Running SwiftFormat...
Reading config file at /Users/.../swift-package-manager/.swiftformat
SwiftFormat completed in 1.8s.
449/652 files formatted, 1 file skipped.
```

File: Sources/PackageLoading/PackageBuilder.swift (hunk 10 above)
Comment (dmhts) [reply Sources/PackageLoading/PackageBuilder.swift:535]: `Xcode` has almost pulled `conversion to guard expression` off :D

File: Sources/Workspace/Workspace.swift (hunk 14 above)
Comment (dmhts) [reply Sources/Workspace/Workspace.swift:2095]: Right, on the one hand, it belongs there, but on the other, we would have to provide the artifacts directory each time making the call site a bit more verbose, i.e: `existingArtifact.is(at: self.location.artifactsDirectory)`, would that be fine?

Comment (tomerd) [reply Sources/Workspace/Workspace.swift:2095]: your call!

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@dmhts are there further modification you want to make, or  is this ready to go?

File: Sources/Workspace/Workspace.swift (hunk 12 above)
Comment (tomerd) [reply Sources/Workspace/Workspace.swift:1729]: yea. what I normally do is point it to a single file / directory: `swiftformat "/path/to/your/code/"`

Comment (tomerd) [reply Sources/Workspace/Workspace.swift:1729]: oh but I was reading the code wrong originally, the formatting is actually fine

File: Sources/Workspace/Workspace.swift (hunk 13 above)
Comment (tomerd) [reply Sources/Workspace/Workspace.swift:1763]: this is my only remaining concern ^^

File: Sources/Workspace/Workspace.swift (hunk 12 above)
Comment (dmhts) [reply Sources/Workspace/Workspace.swift:1728]: Okay, then I reverted the change. Running `swiftformat Sources/Workspace/Workspace.swift` also leads to hundreds of changes in the `Workspace` file, it looks like `swiftformat` was introduced a few years later the file was created, and it hasn't been formatted since then.

File: Sources/Workspace/Workspace.swift (hunk 14 above)
Comment (dmhts) [reply Sources/Workspace/Workspace.swift:2094]: Then I would leave it as is :)

File: Sources/Workspace/Workspace.swift (hunk 13 above)
Comment (dmhts) [reply Sources/Workspace/Workspace.swift:1762]: This check makes sure that the existing artifact is removed only if it exists in the artifacts directory, which is never the case for local non-archived artifacts. You're right that requirement has to be covered with a test, [added one](https://github.com/apple/swift-package-manager/pull/3746/commits/554e2990d0ac4242f4086af756ad195ef81744e5).

File: Sources/Workspace/Workspace.swift (hunk 6 above)
Comment (dmhts) [reply Sources/Workspace/Workspace.swift]: [Done](https://github.com/apple/swift-package-manager/pull/3746/commits/338312aaeb293cd970e8d05f9e2400cb5229ac42), I hope the approach with a protocol extension works.

**Comment by dmhts** [comment]
> @dmhts are there further modifications you want to make, or is this ready to go?

I have only 2 things left on my list:
- making the artifact extraction logic DRYer by reusing it in the `download(_ artifacts: ...` function: That might involve quite some refactoring. I am not sure if it is really needed right now since it still follows [the rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) and it remains implementation details which can be refactored when there is a real need.
- updating the public documentation at `PackageDescription.md`: It already mentions `This path can point directly to a binary artifact or to an archive file that contains the binary artifact at its root.` and I am not sure what is really meant by "an archive file" here, but it already conveniently matches the new piece of functionality, so there is no need of change, right?

File: Sources/Workspace/Workspace.swift (hunk 2 above)
Comment (dmhts) [reply Sources/Workspace/Workspace.swift]: Resolving this one, since the discussion is continued [in the comments](https://github.com/apple/swift-package-manager/pull/3746#issuecomment-934372019).

**Comment by tomerd** [comment]
agreed on these two, perhaps addd a TODO comment on the download func if not already there

was the concern about testing we do not delete local dependencies addressed?

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by dmhts** [comment]
>  perhaps addd a TODO comment on the download func if not already there

[done](https://github.com/apple/swift-package-manager/pull/3746/commits/aad1c7cf6e99b6200187213959b8d3665301a9d7)

> was the concern about testing we do not delete local dependencies addressed?

yes, it was addressed [here](https://github.com/apple/swift-package-manager/pull/3746#discussion_r722025164) and [here](https://github.com/apple/swift-package-manager/pull/3746/commits/554e2990d0ac4242f4086af756ad195ef81744e5)

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Review by tomerd — APPROVED** [review]
nice work! will merge when CI is green
